Basic/Invaders/Assets/Scripts/EnemyBullet.cs

```diff
@@ -16,55 +16,74 @@ public class EnemyBullet : MonoBehaviour
     [SerializeField]
     ParticleSystem m_ShieldExplosionParticle;
 
+    void Awake()
+    {
+        enabled = false;
+    }
+
+    public override void OnNetworkSpawn()
+    {
+        enabled = IsServer;
+    }
+
     private void Start()
     {
         Assert.IsTrue(InvadersGame.Singleton);
-        Assert.IsTrue(NetworkManager.Singleton);
 
         if(InvadersGame.Singleton)
             InvadersGame.Singleton.isGameOver.OnValueChanged += OnGameOver;
     }
 
     private void Update()
     {
-        if (!NetworkManager.Singleton.IsServer) return;
+        if (!IsServer) return;
         transform.Translate(0, -m_TravelSpeed * Time.deltaTime, 0);
 
         if (transform.position.y < k_YBoundary) Destroy(gameObject);
     }
 
-    private void OnDestroy()
+    public override void OnDestroy()
     {
+        base.OnDestroy();
         if (InvadersGame.Singleton) InvadersGame.Singleton.isGameOver.OnValueChanged -= OnGameOver;
     }
 
     private void OnTriggerEnter2D(Collider2D collider)
     {
-        if (!NetworkManager.Singleton.IsServer)
+        // several OnTriggerEnter2D calls may be invoked in the same frame (for different Colliders), so we check if
+        // we're spawned to make sure we don't trigger hits for already despawned bullets
+        if (!IsServer || !IsSpawned)
             return;
 
         var hitPlayer = collider.gameObject.GetComponent<PlayerControl>();
         if (hitPlayer != null)
         {
             hitPlayer.HitByBullet();
-            Destroy(gameObject);
+            NetworkObject.Despawn();
             return;
         }
 
         var hitShield = collider.gameObject.GetComponent<Shield>();
         if (hitShield != null)
         {
+            SpawnExplosionVFXClientRpc(transform.position, Quaternion.identity);
+            
             Destroy(hitShield.gameObject);
-            Destroy(gameObject);
-            Instantiate(m_ShieldExplosionParticle, transform.position, quaternion.identity);
+            NetworkObject.Despawn();
         }
     }
+    
+    [ClientRpc]
+    void SpawnExplosionVFXClientRpc(Vector3 spawnPosition, Quaternion spawnRotation)
+    {
+        Instantiate(m_ShieldExplosionParticle, spawnPosition, spawnRotation);
+    }
 
     private void OnGameOver(bool oldValue, bool newValue)
     {
         enabled = false;
 
         // On game over destroy the bullets
-        if (NetworkManager.Singleton.IsServer) Destroy(gameObject);
+        if (IsServer) NetworkObject.Despawn();
```

Comment:
If moving the OnGameOver subscription to OnNetworkSpawn, we could make sure to only do it on the server so we would no longer need this IsServer check here either